### ISSUE TITLE
What's difference between pad_prob functions and custom gstreamer plugin

### ISSUE THREAD

**Issue by PythonImageDeveloper** [openvinotoolkit/dlstreamer_gst#134]
Hi,
Suppose I want to modify the meta data and doing somethings on buffer or metadata.
**Q1-** Is it possible to modify the metadata inside the prob functions?

**Q2-** As you know, we can access to buffer/metadata inside prob functions or custom plugin, and prob functions hold the buffer until return PadProb.OK like somethings, so we can't to do complex processing inside the prob functions,The pipeline is locked.  I want to know, this restriction also true for custom plugins?

**Comment by nnshah1** [comment]
Q1: Yes definitely. The gvapython / c++ examples demonstrate how to modify buffer and metadata. Let us know if you have a specific issue or concern.

Q2: custom plugins generally also block until they process the frame. However, they can schedule work on different threads and one can place a "queue" element in front of another element to make the left side of the pipeline unblocked from the right side. 
One quick way to prototype a custom element that doesn't transform the caps of a frame is to to use the `gvapython` element (as you can also place a queue in front of it).

**Comment by PythonImageDeveloper** [comment]
@nnshah1 , Thanks a lot for quickly and nicely explain. I appreciate you.

**Q1-** I think when we add elements to gstreamer pipeline, the GStreamer framework execute each of elements as async, and I saw in the gstreamer docs mention that framework is async framework, So therefore each of element to work async I need to add queue element in front of each element? What's cons of queue?

**Q2-** Suppose I create custom plugin for detection/Classification or any task of deep model and I initializing the model is of and in the **do_transform_ip(buffer)** method of gstreamer, I want to feed the buffer to model, One way is to convert buffer to numpy array then pass to model and I think this way is not efficient way, **I want to know, I have to convert buffer to np.array or have another ways for doing this?** What's cons of this converting occurred in embedded devices that has share memory?
I used decoder and the buffer is NV12 format.

**Comment by mliu2020** [comment]
@PythonImageDeveloper,

Sorry for the late response.

**Q1-** The prob function can't block process?

> As @nnshah1 explained, the prob function(also the callback function on pad) locks the video frame and its meta data until it is completed, but if there is a queue element in front of the element, the element in front of the queue could be unblock by running the follow up process in other process.

**Q2-** I thought when we add elements to gstreamer pipeline, the GStreamer framework execute each of elements as async(unblock), and I saw in the gstreamer docs mention that framework is async framework, So therefore each of element to work async I need to add queue element in front of each element? What's cons of queue?

> Yes, the whole GStreamer framework can be run asynchronized, but for each frame, the process order must follow the order defined by pipeline, this is guaranteed by the internal frame buffer queue.

**Q3-** Suppose I write custom plugin for detection/Classification or any task of deep model and I initializing the model and in the **do_transform_ip(buffer)** method of gstreamer, I want to feed the buffer to model, One way is to convert buffer to numpy array then pass to model and I think this way is not efficient way, **I want to know, I have to convert buffer to np.array or have another ways for doing this?** What's cons of this converting occurred in embedded devices that has share memory?

> Could you be more clarify your question? If it is memory sharing between decoder and OpenVINO, there is no solution yet. 

> Intel OpenVINO IE defines the data buffer it accepts, normally we use OpenCV buffer. To do the “converting”, you have to copy the data from the frame buffer which causes CPU cycles. For example, in DL Streamer Python, we can construct numpy.ndarray using the GstBuffer, and then pass this ndarray to IE inference.

Mark

**Comment by mliu2020** [comment]
@PythonImageDeveloper,

Do you have more questions?

If no more question, I want to close this issue in a few days.

Mark Liu

**Comment by mliu2020** [comment]
@PythonImageDeveloper,

Since I don't receive your reply, I assume you might already resolve this issue.

Let me close this issue for now so we can move forward. But feel free to re-open it or post a new issue against our latest release.

Thanks so much for using our product.

Mark Liu
